### PR TITLE
feat: API endpoint for repository permissions in connected services

### PR DESCRIPTION
**Pull request by leafty**
Add new API endpoint to let users check git repository permissions with connected services.

* `/oauth2/api/repositories/<repository_url>` can be used to check `pull`/`push` permissions
* `/oauth2/api/repositories/<repository_url>/probe` can be used to check if a repository is publicly accessible

/deploy #notest renku=master renku-ui=leafty/repository-permissions

See also: SwissDataScienceCenter/renku-ui#3136

### CODE REVIEW

File: components/renku_data_services/migrations/versions/788840ccdcf4_add_connected_services.py [hunk 1]
```diff
@@ -30,7 +30,7 @@ def upgrade() -> None:
         sa.Column("kind", sa.Enum("gitlab", "github", name="providerkind"), nullable=False),
         sa.Column("scope", sa.String(), nullable=False),
         sa.Column("url", sa.String(), nullable=False),
-        sa.Column("client_secret", sa.String(length=500), nullable=True),
+        sa.Column('client_secret', sa.LargeBinary(), nullable=True),
```
Comment (Panaetius) [inline components/renku_data_services/migrations/versions/788840ccdcf4_add_connected_services.py:33]: Formatting error? should be double quotes

File: test/components/renku_data_services/connected_services/test_encryption.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+from renku_data_services.app_config import Config
+from renku_data_services.base_models import APIUser
+from renku_data_services.connected_services import apispec
+from renku_data_services.connected_services import orm as schemas
+from renku_data_services.utils.cryptography import decrypt_string
+
+
+@pytest.fixture
+def admin_api_user() -> APIUser:
```
Comment (Panaetius) [inline test/components/renku_data_services/connected_services/test_encryption.py:17]: we already have an admin_user fixture in conftest that can be reused

**Review by Panaetius — CHANGES_REQUESTED** [review]
looks all good, just two minor comments

**Review by olevski — CHANGES_REQUESTED** [review]
Thanks for implementing this Flora.

As I mentioned in the comments below and when we discussed. I dont think we should mix the oauth functionality with the git/repository functionality, and we should separate the git/repository stuff into its own module. I think this can be done by just moving some of the methods and models you have here in the `components/renku_data_services/git` module.

File: components/renku_data_services/connected_services/external_models.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
+"""Models for external API called by connected services."""
+
+from typing import ClassVar
+
+from pydantic import BaseModel
+
+from renku_data_services.connected_services import models
+
+
+class GitLabConnectedAccount(BaseModel):
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/connected_services/external_models.py:10]: This is not an internal model, I use validation to validate the json response from GitLab API.

File: components/renku_data_services/connected_services/utils.py [hunk 4]
```diff
@@ -3,8 +3,20 @@
...
 def generate_code_verifier(size: int = 48) -> str:
     """Returns a randomly generated code for use in PKCE."""
     rand = random.SystemRandom()
     return base64.b64encode(rand.randbytes(size)).decode()
+
+
+async def probe_repository(repository_url: str) -> bool:
+    """Probe a repository to check if it is publicly available."""
+    async with httpx.AsyncClient() as client:
+        url = f"{repository_url}/info/refs?service=git-upload-pack"
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/connected_services/utils.py:18]: This works for HTTP git repositories. It follows the http protocol for git.

I don't think we can support the ssh protocol for now.

File: components/renku_data_services/connected_services/utils.py [hunk 5]
```diff
@@ -3,8 +3,20 @@
...
+import httpx
+
 
 def generate_code_verifier(size: int = 48) -> str:
     """Returns a randomly generated code for use in PKCE."""
     rand = random.SystemRandom()
     return base64.b64encode(rand.randbytes(size)).decode()
+
+
+async def probe_repository(repository_url: str) -> bool:
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/connected_services/utils.py:15]: I don't think a test should hit the network.

[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/connected_services/utils.py:15]: * the wider internet, as opposed to local dev container services

**Comment by leafty** [comment]
@olevski this is an important feature for Renku 2.0, I do not think there is dependency for short-term work on this, but it is still a core feature of Renku 2.0.

File: components/renku_data_services/connected_services/api.spec.yaml [hunk 6]
```diff
@@ -189,6 +189,47 @@ paths:
...
+            application/json:
+              schema:
+                $ref: "#/components/schemas/RepositoryProviderMatch"
+        "404":
+          description: There is no available provider for this repository.
+        default:
+          $ref: "#/components/responses/Error"
+      tags:
+        - oauth2
+  /oauth2/api/repositories/{repository_url}/probe:
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/connected_services/api.spec.yaml:215]: This is not really how this works. The first endpoint only supports git repositories hosted by one of the listed providers. This one is about checking if the repo is likely clonable by checking if it answers on the git https protocol.

File: components/renku_data_services/connected_services/db.py [hunk 7]
```diff
@@ -317,10 +323,138 @@ async def get_oauth2_connection_token(self, connection_id: str, user: base_model
             token_model = models.OAuth2TokenSet.from_dict(oauth2_client.token)
             return token_model
 
+    async def get_repository(
+        self,
+        repository_url: str,
+        user: base_models.APIUser,
+        etag: str | None,
+        get_internal_gitlab_user: Callable[..., Coroutine[Any, Any, base_models.APIUser]],
+    ) -> models.RepositoryProviderMatch | Literal["304"]:
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/connected_services/db.py:332]: This is not an easy decoupling, as the second part of the code uses the Authlib client configured from the first part.

File: components/renku_data_services/connected_services/api.spec.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (olevski) [reply components/renku_data_services/connected_services/api.spec.yaml:215]: I see. Can you please clarify this in the docstrings of the methods for these endpoints in the blueprint?

File: components/renku_data_services/connected_services/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (olevski) [reply components/renku_data_services/connected_services/utils.py:18]: There is no validation for the URL of a repo when a project is created in the backend. So we should add that in. I think you can specify a format in the apispec and that will generate the right validation for pydantic. I think the ui expects a valid url, in testing when I added a SSH url in the CI deployment the ui crashed.

**Comment by olevski** [comment]
@leafty we can do the refactoring I mentioned at a later time. If you resolve the rest of the comments we can merge this.

File: components/renku_data_services/connected_services/api.spec.yaml [hunk 8]
```diff
@@ -189,6 +189,47 @@ paths:
           $ref: "#/components/responses/Error"
       tags:
         - oauth2
+  /oauth2/api/repositories/{repository_url}:
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/connected_services/api.spec.yaml]: Done in 429765aaa8e4fe29db552afbdb91d4eb38d10ba8.

File: components/renku_data_services/connected_services/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/connected_services/utils.py]: What I mean to say is: we cannot probe `git+ssh` repositories (for now). I would avoid adding support for that if it means adding a SSH or git module to the code base.

They may be clonable or not, we don't know. The UI will be fixed.

File: components/renku_data_services/connected_services/external_models.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+                if self.permissions is not None
+                else default_permissions or models.RepositoryPermissions.default()
+            ),
+        )
+
+
+class GitHubRepositoryPermissions(BaseModel):
+    """Repository permissions from a GitHub provider."""
+
+    PUBLIC_VISIBILITY: ClassVar[str] = "public"
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/connected_services/external_models.py]: Done in 429765aaa8e4fe29db552afbdb91d4eb38d10ba8.

File: components/renku_data_services/connected_services/external_models.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    access_level: int
+
+
+class GitLabRepositoryPermissions(BaseModel):
+    """Repository permissions from a GitLab provider."""
+
+    GUEST_ACCESS_LEVEL: ClassVar[int] = 10
+    DEVELOPER_ACCESS_LEVEL: ClassVar[int] = 30
+    PUBLIC_VISIBILITY: ClassVar[str] = "public"
+    INTERNAL_VISIBILITY: ClassVar[str] = "internal"
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/connected_services/external_models.py]: Done in 429765aaa8e4fe29db552afbdb91d4eb38d10ba8.

File: components/renku_data_services/connected_services/dummy_async_oauth2_client.py [hunk 11]
```diff
@@ -23,6 +23,11 @@ async def get(self, url: str, *args: list, **kwargs: dict) -> Response:
         if parsed.path == "/api/v4/user":
             return self._get_account_response()
 
+        print(f"URL = {url}")
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/connected_services/dummy_async_oauth2_client.py]: Done in 429765aaa8e4fe29db552afbdb91d4eb38d10ba8.

File: components/renku_data_services/connected_services/blueprints.py [hunk 12]
```diff
@@ -180,3 +184,55 @@ async def _get_token(_: Request, user: base_models.APIUser, connection_id: str)
             return json(token.dump_for_api())
 
         return "/oauth2/connections/<connection_id>/token", ["GET"], _get_token
+
+    def get_one_repository(self) -> BlueprintFactoryResponse:
```
[the remark this answers is not in the document]
Comment (leafty) [reply components/renku_data_services/connected_services/blueprints.py]: Done in 429765aaa8e4fe29db552afbdb91d4eb38d10ba8.

**Comment by leafty** [comment]
@olevski comments should be addressed now.

**Review by olevski — APPROVED** [review]

**Pull request merged**